avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastSerializerGenerator.java

```diff
@@ -326,21 +304,21 @@ private void processEnum(Schema enumSchema, JExpression enumValueExpression, JBl
          * Register/retrieve the corresponding {@link org.apache.avro.Schema.EnumSchema} from the mapping.
          */
         long enumSchemaFingerprint = Utils.getSchemaFingerprint(enumSchema);
-        if (enumSchemaIdSet.contains(enumSchemaFingerprint)) {
+        if (enumSchemaVarMap.containsKey(enumSchemaFingerprint)) {
           valueToWrite = JExpr.invoke(
-              enumSchemaMapField.invoke("get").arg(JExpr.lit(enumSchemaFingerprint)),
+              enumSchemaVarMap.get(enumSchemaFingerprint),
               "getEnumOrdinal"
           ).arg(enumValueCasted.invoke("toString"));
         } else {
-          enumSchemaIdSet.add(enumSchemaFingerprint);
-          JVar enumSchemaVar = body.decl(codeModel.ref(Schema.class),
+          /**
+           * In Avro-1.4, there is no way to infer/extract enum schema from {@link org.apache.avro.generic.GenericData.EnumSymbol},
+           * so the serializer needs to record the schema id and the corresponding {@link org.apache.avro.Schema.EnumSchema},
+           * and maintain a mapping between the schema id and EnumSchema JVar for future use.
+           */
+          JVar enumSchemaVar = serializerClass.field(JMod.PRIVATE | JMod.FINAL, Schema.class,
               getVariableName(enumSchema.getName() + "EnumSchema"),
-              enumSchemaMapField.invoke("get").arg(JExpr.lit(enumSchemaFingerprint)));
-          JConditional schemaCheckCond = body._if(JExpr._null().eq(enumSchemaVar));
-          JBlock thenBody = schemaCheckCond._then();
-          thenBody.assign(enumSchemaVar, codeModel.ref(Schema.class).staticInvoke("parse").arg(enumSchema.toString()));
-          thenBody.invoke(enumSchemaMapField, "put").arg(JExpr.lit(enumSchemaFingerprint)).arg(enumSchemaVar);
-
+              codeModel.ref(Schema.class).staticInvoke("parse").arg(enumSchema.toString()));
+          enumSchemaVarMap.put(enumSchemaFingerprint, enumSchemaVar);
           valueToWrite = JExpr.invoke(enumSchemaVar, "getEnumOrdinal").arg(enumValueCasted.invoke("toString"));
```

Comment:
I find this code to be a bit spaghetti... could we instead follow a pattern of 

```
JVar enumSchemaVar = enumSchemaVarMap.computeIfAbsent(enumSchemaFingerprint, () -> 
  serializerClass.field(
              JMod.PRIVATE | JMod.FINAL, 
              Schema.class,
              getVariableName(enumSchema.getName() + "EnumSchema"),
              codeModel.ref(Schema.class).staticInvoke("parse").arg(enumSchema.toString()))
);
valueToWrite = JExpr.invoke(enumSchemaVar, "getEnumOrdinal").arg(enumValueCasted.invoke("toString"));
```

... or something like that? Otherwise, the `JExpr.invoke` part duplicated twice, and since it is meta-code, that makes things extra confusing :D ...

Reply:
Thanks for the suggestion. Fixed.

Reply:
Beautiful! Love the red tint :D !